### PR TITLE
The app doesn't display notification window When app is closed

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] ~Did you add unit tests if this modifies the Windows code?~
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

**Repro Steps**

1. Launch Unity iOS/Android test demo app, enable push service.
2. Then close the app completely.
3. Send a push notification from portal.
4. When the notification is received on the device, press it to open the app.
5. Observe whether the app displays a window with the push notification information.

## Related PRs or issues

[AB#65076](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/65076)

### CODE REVIEW

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs [hunk 1]
```diff
@@ -1,67 +1,80 @@
...
+        public static void PrepareEventHandlers()
+        {
+            AppCenterBehavior.InitializingServices += Initialize;
+            AppCenterBehavior.InitializedAppCenterAndServices += PostInitialize;
+        }
+
+        private static void Initialize()
+        {
+            _unityListener.CallStatic("setListener", new PushDelegate());
+            IsAppCenterStart = true;
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs]: @AnastasiaKubova Could you please add some comments about this logic?

**Review by MatkovIvan — DISMISSED** [review]

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs [hunk 2]
```diff
@@ -1,67 +1,85 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AppCenter.Unity.Push.Internal
+{
+    class PushInternal
+    {
+        private static AndroidJavaClass _push = new AndroidJavaClass("com.microsoft.appcenter.push.Push");
+        private static AndroidJavaClass _unityListener = new AndroidJavaClass("com.microsoft.appcenter.pushdelegate.UnityAppCenterPushDelegate");
+        private static bool IsAppCenterStart = false;
```
Comment (annakocheshkova) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs]: IsAppCenterStarted

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs [hunk 3]
```diff
@@ -1,67 +1,85 @@
...
+        }
+
+        private static void Initialize()
+        {
+            _unityListener.CallStatic("setListener", new PushDelegate());
+            IsAppCenterStart = true;
+            
+            // If `ReplayUnprocessedPushNotifications` was called before App Center start 
+            // than need to call it again after App Center was started.
+            if (IsWaitingToReply)
```
Comment (annakocheshkova) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs]: IsWaitingToReplay

**Review by guperrot — COMMENTED** [review]
I see only variables renaming, where is the fix?

EDIT: nvm, github cache issue

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/UWP/PushInternal.cs [hunk 4]
```diff
@@ -1,124 +1,142 @@
...
+
+        public static void EnableFirebaseAnalytics()
+        {
+        }
+
+        internal static void ReplayUnprocessedPushNotifications()
+        {
+            // Verify that the App Center was started, otherwise set a flag 
+            // that needs call `ReplayUnprocessedPushNotifications` after the App Center will be started.
+            if(!IsAppCenterStarted)
```
Comment (guperrot) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/UWP/PushInternal.cs]: Please reformat.

**Comment by guperrot** [comment]
So why do we also replay notifications when subscribing in Shared/Push line 33, aren't we possibly emitting notification multiple times now?

**Comment by MatkovIvan** [comment]
@guperrot No, there is a check for it. It just adds replay when we subscribe before initialization, so it cannot be replayed correctly.
I don't see the cases when the callback will be called multiple times

**Review by MatkovIvan — DISMISSED** [review]

**Review by jaeklim — COMMENTED** [review]
Why do you need an execution permission for files?

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs [hunk 5]
```diff
@@ -1,67 +1,85 @@
...
+            AppCenterBehavior.InitializedAppCenterAndServices += PostInitialize;
+        }
+
+        private static void Initialize()
+        {
+            _unityListener.CallStatic("setListener", new PushDelegate());
+            IsAppCenterStarted = true;
+            
+            // If `ReplayUnprocessedPushNotifications` was called before App Center start 
+            // than need to call it again after App Center was started.
```
Comment (jaeklim) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs]:
```suggestion
            // then need to call it again after App Center was started.
```

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs [hunk 6]
```diff
@@ -1,67 +1,85 @@
...
+        public static void PrepareEventHandlers()
+        {
+            AppCenterBehavior.InitializingServices += Initialize;
+            AppCenterBehavior.InitializedAppCenterAndServices += PostInitialize;
+        }
+
+        private static void Initialize()
+        {
+            _unityListener.CallStatic("setListener", new PushDelegate());
+            IsAppCenterStarted = true;
```
Comment (jaeklim) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs]: Same code is repeated. Maybe you can move those to Shared/Push and call a method from here. Not a big deal though.

File: CHANGELOG.md [hunk 7]
```diff
@@ -19,6 +19,8 @@ Updated native SDK versions:
 
 ### App Center Push
 
+* **[Bug fix]** Fixed receiving push notifications from background.
```
Comment (jaeklim) [inline CHANGELOG.md]:
```suggestion
* **[Bug fix]** Fix receiving push notifications from background.
```

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs (hunk 6 above)
Comment (MatkovIvan) [reply Assets/AppCenter/Plugins/AppCenterSDK/Push/Android/PushInternal.cs]: The idea is to replay it when platform specific listener is initialized, so it looks OK to have this check in platform specific initialize

**Comment by guperrot** [comment]
> @guperrot No, there is a check for it. It just adds replay when we subscribe before initialization, so it cannot be replayed correctly.
> I don't see the cases when the callback will be called multiple times

Why do we need the shared Push.cs event replay code in that case?
What are the different scenarios here? This is really confusing.

**Comment by MatkovIvan** [comment]
@guperrot this is order of initialization issue. We should deliver the event once all things are initialized

File: CHANGELOG.md [hunk 8]
```diff
@@ -19,6 +19,8 @@ Updated native SDK versions:
 
 ### App Center Push
 
+* **[Bug fix]** Fix receiving push notifications from background.
```
Comment (guperrot) [inline CHANGELOG.md]: This needs to be move to a new version (might need to catch up with release changes before).

**Review by guperrot — CHANGES_REQUESTED** [review]

File: Assets/AppCenter/Plugins/AppCenterSDK/Push/Shared/Push.cs [hunk 9]
```diff
@@ -1,98 +1,124 @@
...
+            // Make a copy of the event so that it isn't suddenly null at the time
+            // of invoking
+            var eventCopy = _pushNotificationReceived;
+            if (eventCopy != null)
+            {
+                eventCopy.Invoke(null, e);
+            }
+        }
+
+        public static void ReplyPushNotificationsIfWatting()
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Push/Shared/Push.cs]: Typo

Comment (annakocheshkova) [reply Assets/AppCenter/Plugins/AppCenterSDK/Push/Shared/Push.cs]: even 2 typos

**Comment by MatkovIvan** [comment]
(do not merge label - because re-testing after the last change is in progress)

**Review by AnastasiaKubova — COMMENTED** [review]
re-testing has done, it works as expected.

**Review by annakocheshkova — DISMISSED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -22,8 +30,6 @@ Updated native SDK versions:
 
 * **[Behavior change]** Fix a security issue with the `Exception` field on `ErrorReport` objects. As a result, the `Exception.StackTrace` now holds the raw stack trace, and the `Exception.Message` field is `null`.
 
-### App Center Push
-
```
Comment (MatkovIvan) [inline CHANGELOG.md]: Revert it

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,5 +1,13 @@
 # App Center SDK for Unity Change Log
 
+## Release 2.6.0 (Under development)
+
+### App Center Push
+
+* **[Bug fix]** Fix receiving push notifications from background.
```
Comment (MatkovIvan) [inline CHANGELOG.md]: Only calling the event has been fixed. Receiving worked before the PR

**Review by guperrot — APPROVED** [review]

**Pull request merged**